### ISSUE TITLE
Targeting NET Standard

### ISSUE THREAD

**Issue by snobu** [projectkudu/ARMClient#30]
Judging by the output of [ApiPort](https://github.com/Microsoft/dotnet-apiport)
it may or may not be a big effort to target NET Standard instead of full NET Framework :)

Given ARMClient is such a handy thing to have around i wonder if we could port this over so it runs on NET Core cross-plat.

![image](https://cloud.githubusercontent.com/assets/6472374/22780735/69d99762-eec8-11e6-8fdb-bedd9bac197c.png)

Currently it doesn't seem to work on Mono - complains about missing Windows Forms when trying to authenticate. Although it MIGHT work if you manage to produce the token cache some other way.

Just a thought..

**Comment by davidebbo** [comment]
Yes, would be great to have it work on Core. For auth, I think we'd need to switch to a model like what xplat CLI is doing. It gives you some code that you have to paste in a browser page, and then auth continues without needed to ever pop up UI. It's just too hard to pop up a browser window in a portable way.

Here is what it looks like with cli, though I don't know about implementation:

```
info:    Executing command login
info:    To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code EKCCKBJ6D to authenticate.
```

**Comment by snobu** [comment]
I saw that in the Mac PowerShell and [**az**](https://github.com/Azure/azure-cli) as well, so it must be coming from ADAL. I'll do some detective work since PowerShell is now open. This magic must be revealed.

**Comment by snobu** [comment]
Alright, so it's called [OAuth 2.0 Device Flow Grant](https://alexbilbie.com/2016/04/oauth-2-device-flow-grant/), it's in ADAL and it goes like this:

https://github.com/Azure-Samples/active-directory-dotnet-deviceprofile/blob/master/DirSearcherClient/Program.cs#L205-L209



    DeviceCodeResult codeResult = ctx.AcquireDeviceCodeAsync(resource, clientId).Result;
    Console.ResetColor();
    Console.WriteLine("You need to sign in.");
    Console.WriteLine("Message: " + codeResult.Message + "\n");
    result = ctx.AcquireTokenByDeviceCodeAsync(codeResult).Result;

**Comment by davidebbo** [comment]
Ha yes, now that I hear 'device flow', that rings a bell from some thread I was on when they were coming up with it for xplat CLI. If you can make it work here, that's awesome.

And hopefully that's the only thing preventing it from using Core (or Mono). The rest should be very portal http request logic.

**Comment by snobu** [comment]
Well, Device Flow login now works ([build this branch](https://github.com/projectkudu/ARMClient/tree/snobudeviceflow)) but i broke the X509 cert auth in the process and i'm not sure how to get out of this one:
![image](https://cloud.githubusercontent.com/assets/6472374/23029324/c7def870-f472-11e6-8bcb-ae71824dcc5e.png)

```
C:\lab\armclient\bin\Debug>armclient devicelogin
Using OAuth 2.0 Device Flow
You need to sign in.
Message: To sign in, use a web browser to open the page https://aka.ms/devicelogin and
enter the code DY2RRGXVN to authenticate.

Got Bearer token from Device Flow: eyJ0eXAiOiJKV1QiLCJh..
User: xxxxxxxxx@microsoft.com, Tenant: xxxxxxxxxxxxxxxx (microsoft.onmicrosoft.com)
        There are 3 subscriptions
        Subscription 6xxxxxxxxxxxxxxxx (yyyyyyyyyy)
        Subscription axxxxxxxxxxxxx (zzzzzzzzzzzzzzzz)
        Subscription bxxxxxxxxxxxxxxx (qqqqqqqqqqqqqq)

User: xxxxxxxxxx@microsoft.com, Tenant: xxxxxxxxxxxxxxxxxxxx (swearjarbank.onmicrosoft.com)
        There are 0 subscriptions


C:\lab\armclient\bin\Debug>armclient get %testurl%
{
  "id": "/subscriptions/axxxx-xxxx-xxxxx-xxxxxx/resourceGroups/securitywhat",
  "name": "securitywhat",
  "location": "eastus",
  "properties": {
    "provisioningState": "Succeeded"
  }
}

C:\lab\armclient\bin\Debug>armclient listcache
User: xxxxxx@microsoft.com, Tenant: xxxxx-xxxxx-xxxxxx-xxxxxx (microsoft.onmicrosoft.com)
        There are 3 subscriptions
        Subscription xxxxxxxx
        Subscription yyyyyyy
        Subscription zzzzzzzz

User: xxxxx@microsoft.com, Tenant: xxxx-xxxx-xxxx-xxxx (swearjarbank.onmicrosoft.com)
        There are 0 subscriptions
```

Not sure about:
- Token store and refresh tokens, since ADAL 3.0 no longer exposes refresh tokens. Seems the token gets stored correctly in `%userprofile%\.arm`, but will it work in 1 hour? Didn't test.
- I'm using `common` [here ](https://github.com/projectkudu/ARMClient/blob/snobudeviceflow/ARMClient.Authentication/AADAuthentication/BaseAuthHelper.cs#L186) as tenant, am i doing it right? The token seems to be fine.

I'm still testing but would value your input here.

**Comment by snobu** [comment]
Looks like i broke the token refresh mechanism.
This is what happens after >1h from devicelogin:

```
C:\lab\armclient\bin\Debug>armclient get "https://management.azure.com/subscriptions/x-x-x-x/resourceGroups/?api-version=2014-04-01"

The method or operation is not implemented.
```

Well, back to the drawing board.

![back](https://cloud.githubusercontent.com/assets/6472374/23056853/04beea74-f4f6-11e6-9ac7-b08567db3672.png)

**Comment by davidebbo** [comment]
Presumably, the xplat CLI is able to continue to work after one hour? Sorry, I'm not very familiar with the refresh mechanism.

**Comment by snobu** [comment]
Yes it is. I suspect i'm not saving the correct thing to offline cache with ADAL 3. I'll figure it out.

**Comment by fawohlsc** [comment]
Any plans to make armclient cross-platform by porting to .NET Core? Would be handy to have it in Cloud Shell, Linux and Mac.

**Comment by snobu** [comment]
I believe this effort is now abandoned in favor of Azure CLI's `az rest` which is cross-platform from birth.

Example -

```
$ az rest --method get --url https://management.azure.com/subscriptions/<SUBSCRIPTION_ID>?api-version=2021-01-01

{
  "authorizationSource": "Legacy",
  "displayName": "xx",
  "id": "/subscriptions/xxxxxxxx",
  "managedByTenants": [
    {
      "tenantId": "xxxxxxx"
    },
    {
      "tenantId": "xxxxxxx"
    }
  ],
  "state": "Enabled",
  [...]
```

**Comment by fawohlsc** [comment]
Thanks @snobu!
